subscription_oca/models/sale_subscription.py

```diff
@@ -0,0 +1,807 @@
+                        composition_mode="comment",
+                        email_layout_xmlid="mail.mail_notification_paynow",
+                    )
+                invoice_number = invoice_id.name
+                message_body = (
+                    "<b>%s</b> <a href=# data-oe-model=account.move data-oe-id=%d>%s</a>"
+                    % (msg_static, invoice_id.id, invoice_number)
+                )
+
+        if self.template_id.invoicing_mode == "sale_and_invoice":
+            order_id = self.create_sale_order()
+            order_id.action_done()
+            new_invoice_id = order_id._create_invoices()
+            new_invoice_id.action_post()
+            new_invoice_id.invoice_origin = order_id.name + ", " + self.name
+            invoice_number = new_invoice_id.name
+            message_body = (
+                "<b>%s</b> <a href=# data-oe-model=account.move data-oe-id=%d>%s</a>"
+                % (msg_static, new_invoice_id.id, invoice_number)
+            )
+        if not invoice_number:
+            invoice_number = _("To validate")
+            message_body = "<b>%s</b> %s" % (msg_static, invoice_number)
+
+        self.message_post(body=message_body)
+
+    def manual_invoice(self):
+        invoice_id = self.create_invoice()
+        self._onchange_template_id()
+        context = dict(self.env.context)
+        context["form_view_initial_mode"] = "edit"
+        return {
+            "name": self.name,
+            "views": [
+                (self.env.ref("account.view_move_form").id, "form"),
+                (self.env.ref("account.view_move_tree").id, "tree"),
+            ],
+            "view_type": "form",
+            "view_mode": "form",
+            "res_model": "account.move",
+            "res_id": invoice_id.id,
+            "type": "ir.actions.act_window",
+            "context": context,
+        }
+
+    @api.depends("invoice_ids", "sale_order_ids.invoice_ids")
+    def _compute_account_invoice_ids_count(self):
+        for record in self:
+            record.account_invoice_ids_count = len(self.invoice_ids) + len(
+                self.sale_order_ids.invoice_ids
+            )
+
+    def action_view_account_invoice_ids(self):
+        return {
+            "name": self.name,
+            "views": [
+                (self.env.ref("account.view_move_tree").id, "tree"),
+                (self.env.ref("account.view_move_form").id, "form"),
+            ],
+            "view_type": "form",
+            "view_mode": "tree,form",
+            "res_model": "account.move",
+            "type": "ir.actions.act_window",
+            "domain": [
+                ("id", "in", self.invoice_ids.ids + self.sale_order_ids.invoice_ids.ids)
+            ],
+            "context": self.env.context,
+        }
+
+    @api.depends("sale_order_ids")
+    def _compute_sale_order_ids_count(self):
+        for record in self:
+            record.sale_order_ids_count = len(self.sale_order_ids)
```

Comment:
watch out w/ all these use of `len(rec.an_x2many)` it can be very slow.
What about using a read_group? Is a common approach, you can find several examples in the OCA and odoo code too.